token/src/token.rs

```diff
@@ -0,0 +1,89 @@
+use bulletproofs::{BulletproofGens, PedersenGens};
+use curve25519_dalek::scalar::Scalar;
+use zkvm::*;
+
+pub struct Token {
+    flavor: Predicate,
+    metadata: Vec<u8>,
+    nonce: Scalar,
+}
+
+impl Token {
+    pub fn new(pred: Predicate, metadata: &[u8]) -> Self {
+        Token {
+            flavor: pred,
+            metadata: metadata.to_vec(),
```

Comment:
If you need a `Vec<u8>`, you may simply ask for it right here and move it in w/o extra allocations.

https://rust-lang-nursery.github.io/api-guidelines/flexibility.html

<img width="944" alt="image" src="https://user-images.githubusercontent.com/698/53603090-55479100-3b65-11e9-9f58-53ac8f5c6eca.png">